### ISSUE TITLE
Request details on hyperparameters of the VBD corpus

### ISSUE THREAD

**Issue by CSeIn** [wangchengzhong/GALDSE#1]
Thank you for sharing the code. I'm interested in your work.
I want to train the model in the VBD corpus. 
Could you give me more details (e.g. batch size, epoch, hyperparameters, use_ts_conformer, PESQ & SI-SNR loss weights) for the VBD corpus?

**Comment by wangchengzhong** [comment]
Hello. The default parameters in this codebase work well for training on the VBD dataset:

- use_ts_conformer: True
- PESQ weight: 5e-4
- SI-SNR loss weight: 1e-5
- Training epochs: ~800
- Batch size: 32 (in total)

**Comment by CSeIn** [comment]
Thank you. 
If you experiment with that configuration, what are the PESQ performance scores?
Could you give me the hyperparameters for the noise schedule (\alpha_T, \alpha_1, \kappa, p) in the VBD corpus?

**Comment by wangchengzhong** [comment]
With the default hyperparameters in this updated codebase (kappa=0.3, alpha_T=0.99, alpha_1=0.003), the model achieves a PESQ score of 3.27 on the VBD dataset.

**Comment by wangchengzhong** [comment]
Thanks for your question. Based on training logs, the hyperparameters in this codebase works well for training on the VBD dataset, achieving a PESQ score of 3.254:

- use_ts_conformer: True
- PESQ weight: 5e-4
- SI-SNR loss weight: 1e-5
- Training epochs: ~800-1000
- Batch size: 42 (total)
- alpha_T: 0.99
- alpha_0: 0.003
- kappa: 0.3
- power: 0.39

**Comment by CSeIn** [comment]
Thank you again for your help.

I have been training on the VBD dataset using the configuration you provided. However, I’ve noticed that completing even 1 epoch takes 1 hour (with 4 GPU 3090). I’m wondering if the PESQ loss calculation might be a major factor contributing to slower training times. If you have any insights on why the training might be taking so long or suggestions on how to optimize it, I would greatly appreciate your advice.

Additionally, I would like to ask about the configuration you used to obtain the PESQ score of 3.19 on VBD dataset as reported in your paper. Could you please share the specific hyperparameters and training details?

**Comment by wangchengzhong** [comment]
Thanks.

## Regarding slow training with PESQ loss

The slow training speed you're experiencing is likely related to the PESQ loss calculation. We've encountered similar issues with slow training on certain hardware configurations (4x NVIDIA 3090 + Xeon 4214R), while it performed better on different hardware (Xeon 5318Y + 4x NVIDIA A40).

This is a known issue with `torchaudio.functional.lfilter` running slowly on GPUs but faster on CPUs, as discussed in [this PyTorch forum thread](https://discuss.pytorch.org/t/torchaudio-functional-lfilter-runs-very-slow-on-gpu-but-fast-on-cpu/142013) and [this torch-pesq issue](https://github.com/audiolabs/torch-pesq/issues/9).

To improve training speed, we recommend modifying the `align_level` and `preemphasize` functions in the `PESQLoss` class to move tensors to CPU when performing the `lfilter` operation:


`    
@typechecked
    def align_level(
        self, signal: TensorType["batch", "sample"]
    ) -> TensorType["batch", "sample"]:
        """Align power to 10**7 for band 325 to 3.25kHz

        Parameters
        ----------
        signal : TensorType["batch", "sample"]
            Input time signal with size [batch, sample]

        Returns
        -------
        TensorType["batch", "sample"]
            Tensor containing the scaled time signal
        """
        device = signal.device
        
        cpu_signal = signal.cpu()
        cpu_power_filter = self.power_filter.cpu()
        
        filtered_signal = lfilter(
            cpu_signal, cpu_power_filter[1], cpu_power_filter[0], clamp=False
        )
        
        filtered_signal = filtered_signal.to(device)

        # calculate power with weird bugs in reference implementation
        power = (
            (filtered_signal**2).sum(dim=1, keepdim=True)
            / (filtered_signal.shape[1] + 5120)
            / 1.04684
        )

        signal = signal * (10**7 / power).sqrt()

        return signal

    @typechecked
    def preemphasize(
        self, signal: TensorType["batch", "sample"]
    ) -> TensorType["batch", "sample"]:
        """Pre-empasize a signal

        This pre-emphasize filter is also applied in the reference implementation. The filter
        coefficients are taken from the reference.

        Parameters
        ----------
        signal : TensorType["batch", "sample"]
            Input time signal with size [batch, sample]

        Returns
        -------
        TensorType["batch", "sample"]
            Tensor containing the pre-emphasized signal
        """

        emp = torch.linspace(0, 15, 16, device=signal.device)[1:] / 16.0
        signal[:, :15] *= emp
        signal[:, -15:] *= torch.flip(emp, dims=(0,))
        
        device = signal.device
        
        cpu_signal = signal.cpu()
        cpu_pre_filter = self.pre_filter.cpu()
        
        signal = lfilter(cpu_signal, cpu_pre_filter[1], cpu_pre_filter[0], clamp=False)
        
        signal = signal.to(device)

        return signal
`


With these modifications, we found training speed improved to around 1 second per step on our hardware configuration.
If you still find training too slow, you can use MSE loss for the initial ~800 epochs (which can achieve a PESQ score over 3.10), and then fine-tune with the assisted losses.  The pesq weight is recommended to be a little bit higher when fine-tuning (e.g. 1e-3)

## Regarding the PESQ score on VBD dataset

To reach the performance reported in our paper, you should:

1. Follow the hyperparameters mentioned in the paper
2. Input the guided psm and noisy signal explicitly into the model
3. Fine-tune with the following parameters:
   - spec_factor = 0.15
   - kappa = 0.1
   - spec_abs_exponent = 0.76
   - Other parameters as specified in the paper

However, please note that in our updated codebase, we've found that explicit guidance input in the original training method makes the diffusion model rely too heavily on the guidance signal, which doesn't fully demonstrate the efficiency of our method and may lead to suboptimal results in challenging scenarios. Therefore, we strongly recommend training with the updated codebase.

**Comment by wangchengzhong** [comment]
We're closing this issue since requirements completed
